### PR TITLE
feat(vrl): add `is_empty` function

### PR DESCRIPTION
**Pull request by JeanMertz**
This PR adds an `is_empty` function. It returns `true` if a string or array is empty. We could update it to also allow for objects, but I'm also happy to wait for future contributions to that, if there's ever a need for it.

### CODE REVIEW

**Comment by lucperkins** [comment]
String emptiness is currently covered by the `is_nullish` function (as a subclass of whitespace-only strings). It may be totally fine to have some overlap, just making a note in case that's something we want to avoid.

**Comment by hhromic** [comment]
> We could update it to also allow for objects, (...)

If `is_empty()` will work for strings and arrays, then I bet users will also be expecting for it to "just work" for objects as well, me included. The reasoning is that if I see that `is_empty()` works for different types, I will assume that it actually checks for "emptiness" value of container types. Pretty much like how [Python evalutes truth values for different types](https://docs.python.org/3/library/stdtypes.html#truth-value-testing) (check the third point):

> Here are most of the built-in objects considered false:
> * constants defined to be false: None and False.
> * zero of any numeric type: 0, 0.0, 0j, Decimal(0), Fraction(0, 1)
> * empty sequences and collections: '', (), [], {}, set(), range(0)

So in my humble opinion, missing the test for empty objects leads to an incomplete `is_empty()` implementation.

---

> Yeah, I think the overlap is fine, and `is_nullish` is really something I wouldn't think to search for when I want to know "is this an empty string".

Me neither! :) I didn't even know that function existed before reading this.
After reading its [documentation](https://vector.dev/docs/reference/vrl/functions/#is_nullish) I can fully agree that `is_nullish()` serves a (very) different purpose than `is_empty()`. Just take for example that `is_nullish("-") == true`.

**Review by StephenWakely — CHANGES_REQUESTED** [review]
I think we might as well allow it for objects right away since it's such a simple change and would be expected.

Everything else is looking good!

**Comment by jszwedko** [comment]
I merged in `master` and add support for objects.

File: lib/vrl/stdlib/benches/benches.rs [hunk 1]
```diff
@@ -672,6 +673,35 @@ bench_function! {
...
+    is_empty => vrl_stdlib::IsEmpty;
+
+    empty_array {
+        args: func_args![value: value!([])],
+        want: Ok(true),
+    }
+
+    non_empty_array {
+        args: func_args![value: value!([1, 2, 3])],
+        want: Ok(true),
```
Comment (spencergilbert) [inline lib/vrl/stdlib/benches/benches.rs:686]: false?

File: lib/vrl/stdlib/src/is_empty.rs [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+                result: Ok("true"),
+            },
+            Example {
+                title: "empty object",
+                source: r#"is_empty({})"#,
+                result: Ok("true"),
+            },
+            Example {
+                title: "non-empty array",
+                source: r#"is_empty([null])"#,
```
Comment (spencergilbert) [inline lib/vrl/stdlib/src/is_empty.rs:57]: This feels like a weird example to use with `null` - (but this doesn't get shown to users, just tested?)

**Review by spencergilbert — APPROVED** [review]

File: lib/vrl/stdlib/src/is_empty.rs (hunk 2 above)
Comment (jszwedko) [reply lib/vrl/stdlib/src/is_empty.rs:57]: Yeah, that is weird to use `null` there. It doesn't get shown to users, but I'll update it anyway. This also makes me realize I forgot to update the doc.

File: lib/vrl/stdlib/benches/benches.rs (hunk 1 above)
Comment (jszwedko) [reply lib/vrl/stdlib/benches/benches.rs:686]: Hmm, yeah, I missed that.

File: lib/vrl/stdlib/src/is_empty.rs (hunk 2 above)
Comment (jszwedko) [reply lib/vrl/stdlib/src/is_empty.rs]: Updated in [d55680c](https://github.com/vectordotdev/vector/pull/9732/commits/d55680c0e2ef0db35758bfbc334f2f283e8e3a1d)

File: lib/vrl/stdlib/benches/benches.rs (hunk 1 above)
Comment (jszwedko) [reply lib/vrl/stdlib/benches/benches.rs:686]: Updated in [d55680c](https://github.com/vectordotdev/vector/pull/9732/commits/d55680c0e2ef0db35758bfbc334f2f283e8e3a1d)

Comment (pablosichert) [reply lib/vrl/stdlib/benches/benches.rs:686]: Isn't the assertion still in a mismatch here?

**Review by pablosichert — CHANGES_REQUESTED** [review]

File: lib/vrl/stdlib/benches/benches.rs [hunk 3]
```diff
@@ -672,6 +673,35 @@ bench_function! {
...
+    }
+
+    empty_object {
+        args: func_args![value: value!({})],
+        want: Ok(value!(true)),
```
Comment (pablosichert) [inline lib/vrl/stdlib/benches/benches.rs:691]: Can these be reduced to

```suggestion
        want: Ok(true),
```

as in the other assertions?

File: website/cue/reference/remap/functions/is_empty.cue [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+remap: functions: is_empty: {
+	category: "Type"
+	description: """
+		Check if the object, array, or string is empty.
+		"""
+
+	arguments: [
+		{
+			name:        "value"
+			description: #"The value to check"#
```
Comment (pablosichert) [inline website/cue/reference/remap/functions/is_empty.cue:12]:
```suggestion
			description: #"The value to check."#
```

File: website/cue/reference/remap/functions/is_empty.cue [hunk 5]
```diff
@@ -0,0 +1,49 @@
+package metadata
+
+remap: functions: is_empty: {
+	category: "Type"
+	description: """
+		Check if the object, array, or string is empty.
```
Comment (StephenWakely) [inline website/cue/reference/remap/functions/is_empty.cue:6]:
```suggestion
		Check if the object, array, or string contains any elements.
```

Just thinking using the word empty again in the description is a little redundant given it is already in the name.

**Review by StephenWakely — APPROVED** [review]
Looks good to me!

**Comment by jszwedko** [comment]
Thanks @pablosichert ! I think I made all of the edits if you want to give it another 👀

**Review by pablosichert — APPROVED** [review]
💯
